File: pkg/auth/rights/applications.go
// Copyright © 2018 The Things Network Foundation, The Things Industries B.V.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package rights

import (
	"context"

	"go.thethings.network/lorawan-stack/pkg/errors"
	"go.thethings.network/lorawan-stack/pkg/ttnpb"
	"go.thethings.network/lorawan-stack/pkg/validate"
)

var (
	// ErrPermissionDenied is returned if the rights were insufficient to perform
	// this operation.
	ErrPermissionDenied = &errors.ErrDescriptor{
		MessageFormat: "Permission denied to perform this operation",
		Type:          errors.PermissionDenied,
		Code:          1,
	}

	// ErrInvalidApplicationID is returned if an invalid application ID was passed to an
	// operation that requires it.
	ErrInvalidApplicationID = &errors.ErrDescriptor{
		MessageFormat: "Invalid application ID given",
		Type:          errors.InvalidArgument,
		Code:          2,
	}
)

func init() {
	ErrPermissionDenied.Register()
	ErrInvalidApplicationID.Register()
}

// RequireApplication checks, based on the context, that the clients has the
// specified rights for this application.
func RequireApplication(ctx context.Context, appIdentifiers ApplicationIDGetter, rights ...ttnpb.Right) error {
	// TODO: Accept administrator authorization even if not tied to the application
	// https://github.com/TheThingsIndustries/ttn/issues/731
	if err := validate.ID(appIdentifiers.GetApplicationID()); err != nil {
		return ErrInvalidApplicationID.NewWithCause(nil, err)
	}

	if ad := FromContext(ctx); !ttnpb.IncludesRights(ad, rights...) {
		return ErrPermissionDenied.New(nil)
	}

	return nil
}

File: pkg/webui/console/store/middleware/applications.js
// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { createLogic } from 'redux-logic'

import api from '../../api'
import * as applications from '../actions/applications'
import * as application from '../actions/application'

const getApplicationsLogic = createLogic({
  type: [
    applications.GET_APPS_LIST,
    applications.SEARCH_APPS_LIST,
  ],
  latest: true,
  async process ({ getState, action }, dispatch, done) {
    const { page, pageSize: limit, query } = action.filters
    try {
      const data = query
        ? await api.applications.search({
          page,
          limit,
          id_contains: query,
          name_contains: query,
        })
        : await api.applications.list({ page, limit })
      dispatch(applications.getApplicationsSuccess(data.applications, data.totalCount))
    } catch (error) {
      dispatch(applications.getApplicationsFailure(error))
    }

    done()
  },
})

const getApplicationsRightsLogic = createLogic({
  type: [
    applications.GET_APPS_RIGHTS_LIST,
    application.GET_APP_API_KEY_PAGE_DATA,
    application.GET_APP_COLLABORATOR_PAGE_DATA,
  ],
  async process ({ getState, action }, dispatch, done) {
    const { id } = action
    try {
      const result = await api.rights.applications(id)

      dispatch(applications.getApplicationsRightsListSuccess(result.rights.sort()))
    } catch (error) {
      dispatch(applications.getApplicationsRightsListFailure(error))
    }

    done()
  },
})

export default [
  getApplicationsLogic,
  getApplicationsRightsLogic,
]
